File: insight-brain-frontend/src/main/frontend/utility/services/ProductLicense.js
/*
 * Copyright (c) 2011-present Sonatype, Inc. All rights reserved.
 * Includes the third-party code listed at http://links.sonatype.com/products/clm/attributions.
 * "Sonatype" is a trademark of Sonatype, Inc.
 */
import axios from 'axios';
import { getPermissions } from 'MainRoot/util/authorizationUtil';
import { getLicenseDetailsUrl, getLicenseSummaryUrl } from 'MainRoot/util/CLMLocation';

export default function ProductLicense() {
  return {
    load() {
      return loadIfNotYetLoaded();
    },
  };
}

let licenseInfoPromise = null;
// if licenseLoadedWithoutLoginOrWithError is true re-fetch the license data on next loadIfNotYetLoaded call, the
// license data was loaded while logged out or an error occurred
let licenseLoadedWithoutLoginOrWithError = false;

export function loadIfNotYetLoaded() {
  if (licenseInfoPromise == null || licenseLoadedWithoutLoginOrWithError) {
    licenseInfoPromise = getLicense();
  }
  return licenseInfoPromise;
}

async function getLicense() {
  let hasConfigureSystemPermission = false;

  try {
    hasConfigureSystemPermission = (await getPermissions(['CONFIGURE_SYSTEM'], 'global', 'global', false)).length;
    licenseLoadedWithoutLoginOrWithError = false;
  } catch (error) {
    // Always reload the license on next try if this response is an error
    licenseLoadedWithoutLoginOrWithError = true;
    if (error?.status !== 401) {
      // If it is not a user logged out 401 status, throw the error and let the application handle it
      throw error;
    }
  }

  const licenseUrl = hasConfigureSystemPermission ? getLicenseDetailsUrl() : getLicenseSummaryUrl();
  try {
    return (await axios.get(licenseUrl)).data;
  } catch (error) {
    // Always reload the license on next try if this response is an error
    licenseLoadedWithoutLoginOrWithError = true;
    throw error;
  }
}

// Visible for testing - Clear the loaded (cached) product license promise
export function clearLoadedProductLicensePromise() {
  licenseInfoPromise = null;
}

File: insight-brain-data/src/main/java/com/sonatype/insight/brain/model/configuration/ProductLicense.java
/*
 * Copyright (c) 2011-present Sonatype, Inc. All rights reserved.
 * Includes the third-party code listed at http://links.sonatype.com/products/clm/attributions.
 * "Sonatype" is a trademark of Sonatype, Inc.
 */
package com.sonatype.insight.brain.model.configuration;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;

import com.sonatype.insight.model.HasStringId;

/**
 * @since 1.96
 */
@Entity
@Table(name = "product_license")
public class ProductLicense
    implements HasStringId
{
  @Id
  @Column(name = "product_license_id")
  private String id;

  @Column(name = "license_key")
  private String licenseKey;

  @Column(name = "license_details")
  private String licenseDetails;

  @Override
  public String getId() {
    return id;
  }

  @Override
  public void setId(String id) {
    this.id = id;
  }

  public String getLicenseKey() {
    return licenseKey;
  }

  public void setLicenseKey(String licenseKey) {
    this.licenseKey = licenseKey;
  }

  public String getLicenseDetails() {
    return licenseDetails;
  }

  public void setLicenseDetails(String licenseDetails) {
    this.licenseDetails = licenseDetails;
  }
}
